controllers/src/main/scala/cromwell/pipeline/controller/AkkaHttpClient.scala

```diff
@@ -4,67 +4,113 @@ import akka.actor.ActorSystem
+      }
+    }
+
+  private def positiveResponse[B](
+    response: HttpResponse
+  )(implicit ec: ExecutionContext, f: Format[B]): Future[Response[B]] =
+    if (response.headers.contains(httpContentTypeJson)) {
+      getResponseBody[B](response)(ec, f).map { responseBody =>
+        Response[B](
           response.status.intValue(),
-          body.data.utf8String,
-          response.headers.foldLeft(Map[String, List[String]]()) {
-            case (acc, header) =>
-              acc + (header.name() -> (header.value() :: acc.getOrElse(header.name(), Nil)))
-          }
+          PositiveResponseBody(responseBody.get),
+          parseHeaders(response)
         )
       }
+    } else {
+      Future.failed(ParseJsonException("Could not parse body cause it is not a json"))
+    }
+
+  private def negativeResponse[B](response: HttpResponse)(implicit ec: ExecutionContext): Future[Response[B]] =
+    response.entity.toStrict(expirationTime).map { body =>
+      val pb = parseResponseBody(body)
+      Response[B](
+        response.status.intValue(),
+        NegativeResponseBody(pb.toString()),
+        parseHeaders(response)
+      )
+    }
+
+  private def parseHeaders(response: HttpResponse): Map[String, Seq[String]] =
+    response.headers.foldLeft(Map[String, List[String]]()) {
+      case (acc, header) =>
+        acc + (header.name() -> (header.value() :: acc.getOrElse(header.name(), Nil)))
+    }
+
+  private def getResponseBody[B](
+    response: HttpResponse
+  )(implicit ec: ExecutionContext, f: Format[B]): Future[JsResult[B]] =
+    response.entity.toStrict(expirationTime).map { body =>
+      Json.fromJson[B](parseResponseBody(body))
     }
 
+  private def parseResponseBody(e: HttpEntity.Strict): JsValue =
+    Json.parse(e.data.utf8String)
+
   private def parseHeaders(headers: Map[String, String]): List[RawHeader] =
     headers.map { case (param, value) => RawHeader(param, value) }.toList
 
-  override def post(
+  override def post[B, P](
     url: String,
     params: Map[String, String] = Map(),
     headers: Map[String, String] = Map(),
-    payload: String
-  )(implicit ec: ExecutionContext): Future[Response] = {
+    payload: P
+  )(implicit ec: ExecutionContext, bf: Format[B], pf: Format[P]): Future[Response[B]] = {
     val futureResponse: Future[HttpResponse] = Http().singleRequest(
       HttpRequest(method = HttpMethods.POST, uri = Uri(url).withQuery(Query(params)))
-        .withEntity(payload)
+        .withEntity(Json.stringify(Json.toJson(payload)))
         .withHeaders(parseHeaders(headers))
     )
-    responsify(futureResponse)
+    responsify[B](futureResponse)
   }
 
-  override def put(
+  override def put[B, P](
     url: String,
     params: Map[String, String] = Map(),
     headers: Map[String, String] = Map(),
-    payload: String
-  )(implicit ec: ExecutionContext): Future[Response] = {
+    payload: P
+  )(implicit ec: ExecutionContext, bf: Format[B], pf: Format[P]): Future[Response[B]] = {
```

Comment:
Reads for bf and Writes for pf

Reply:
done, to change it with more concrete parser implementation is a very good idea, it is easier to separate respbody and payload, so thanks